### ISSUE TITLE
RPM to MAS Module Manager mega patch?

### ISSUE THREAD

**Issue by MOARdV** [MOARdV/AvionicsSystems#277]
It'd be a nightmare to write and test.  But it'd make MAS adoption easier.  I just don't see myself having dozens and dozens of hours to write and test a patch like that.

**Comment by MOARdV** [comment]
Specific props that require a second round of work:
- [ ] MAS_DigitalIndicator_Elec_Output has an associated rotary switch, and it selects one of several power outputs.

**Comment by theonegalen** [comment]
Wow, I've just ordered a new computer and you're working on MAS again? It must be my birthday after all!

**Comment by MOARdV** [comment]
> you're working on MAS again?

Well, sort-of working on MAS.  The plugin can do just about everything RPM can do, so it's on par with RPM.  The problem is that I've only created props that I needed for IVAs I created myself, so it's not really well positioned for other people to jump in and create fully-developed IVAs.

The mod's never going to take off if I don't make a patch that lets RPM cockpits work with MAS.  The problem is there are something like 1200 props in the ASET collection, and it is mind-numbingly tedious to go through all of them, figure out what they do, and assemble an equivalent MAS prop, even with the tool I wrote to automate the config file generation step.  But if I don't write a patch to convert RPM IVAs to MAS and provide MAS versions of the RPM props, there's not much of an impetus for people to make the jump from RPM.

I've also discovered quite a few props in the ASET collection that don't have functional config files - there are several different props that Alexustas made that have never shown up in-game.  A few of them I've managed to use in MAS at various points during development, but I'd love to share all of them.  The trick there is that they cover a wide range of design styles - analog and modern (glass) aircraft, Apollo-era spacecraft, more modern spacecraft.  It would be weird to put all of them in one IVA, and my time is a little too limited to create several IVAs to showcase all of them properly.

So, we'll see how many props I can convert this summer before I'm too bored to keep going.  It would be easier if alexustas were still modding, since he figured out how to take advantage of MAS fairly quickly, based on some of the videos he posted on his Youtube channel.  Ah, well.

**Comment by theonegalen** [comment]
Well, now that I have my new computer set up, I'd like to get back to working on IVAs myself. However, I'd hate to duplicate your work unnecessarily, or to have conflicting versions of MAS versions of props. I wish I had the head for MAS code that alexustas has, but the things he was posting on his youtube with the MAS MFDs just completely blew my mind.

**Comment by MOARdV** [comment]
MFDs are a going to be tough - I've got to make MAS versions of what alexustas did, which means figuring out what each page is supposed to be doing.  The rest of the props will be comparatively easy since I've got a tool that takes care of most of the generation work, so all I have to do is add a few lines to an XML file, and a few lines to a MM patch file, and a new prop shows up.  But it still means going through every single prop and figuring out what it does and what the MAS equivalent is.

**Comment by MOARdV** [comment]
alexustas at one point said he was deprecating a number of old switch styles.  While doing this upgrade patch, I should see where I can consolidate them (most likely some of the switchTumble variants).

**Comment by theonegalen** [comment]
Yes, I think most of those can be safely replaced with the modular switch.

**Comment by theonegalen** [comment]
Yes, I'd be glad to get started on the Avionics pack, most likely on Saturday.

**Comment by MOARdV** [comment]
> I'd be glad to get started on the Avionics pack, most likely on Saturday. Do you want me to focus on duplicating the functionality of the RPM versions distributed with the Avionics pack itself?

That is very much appreciated.  My suggestion is to make equivalent props where possible.  The initial goal is to get a replacement for each prop in the parts packs that works the way the player would expect based on their experience with RPM.  If there's things MAS can do to enhance the usefulness of the prop, that's good, too.  For instance, I made the altimeters stop working outside the atmosphere, under the assumption that they use atmospheric pressure to determine altitude, not radar.

The radio nav stuff will be fun to see in use.  alexustas made some neat aviation charts for Kerbin - it'd be nice for them to be useful during play.

Lights or other electrically-controlled components should use fc.Conditioned(), for sure.  For instrument lights, make sure the COLOR_SHIFT uses 'blend = true' - MAS allows instrument lighting to be dimmed, instead of only off/on.  The ASET light controls don't support that, but MAS controls can.

Oh - I should describe the "standards" I'm using for the props.  They're mostly ad-hoc, but I'm trying to be better organized.  Hopefully this looks like more work than it is:

* The MAS prop config file goes under GameData/MOARdV/MAS_ASET/(directory), where `(directory)` is the name of the ASET prop's directory - like ASET_ALTIMETER.  For something like the ASET_B-Knob, they'll all go under ASET_B-Knob, not under ASET_B-Knob/ASET_B-Knob-2pos, ASET_B-Knob/ASET_B-Knob-3pos, etc. - just make sure the prop name includes the 2pos or 3pos, in this example.
* Props are named `MAS_(propname)`, where the (propname) is something similar to the ASET prop's type and name (like MAS_APswitchButtonClearNodes).  Feel free to fix typos in names.  Ideally, all of the prop names in a folder have consistent names, like MAS_LabelText_(name) for all the props in MAS_ASET/LabelText.  I didn't follow that rule myself early on. but I'm trying to be more consistent.
* The config file's name is the same as the prop.  The Prop Config tool does this automatically, but that tool's really only useful if you need to make a lot of props with a similar template.
* Documentation - I've got a work-in-progress document for the entire props collection.  For each folder I convert, I list the folder's name, the general format of the prop names, and a one or two sentence description of what the prop looks like and what it does.  If you've got many props in a folder, you don't have to document each one, as long as the name of the prop provides a hint.  The documentation can just be a text file - I'll merge it into the master document,
* Props that depend on other props (like, maybe, radio navigation?) - these are more complicated, since the props need to share information through persistent variables.  What I really need to do is add a master list of shared persistents to the top post here.  If you come across something like that, let me know, and we'll figure out how to proceed.

(sorry for the wall of text :) )

**Comment by theonegalen** [comment]
Finished the AHI. It's been a long time since I've programmed a prop, so it took a lot of brainpower and time for me to break through my mind fog. I think the others will come more quickly.

I really like how MAS works - I especially like the ability to nest functions in order to create custom variables right in the prop itself. Not having to have a huge CUSTOM_VARIABLES config somewhere is nice.

**Comment by MOARdV** [comment]
> I really like how MAS works - I especially like the ability to nest functions in order to create custom variables right in the prop itself. Not having to have a huge CUSTOM_VARIABLES config somewhere is nice.

That was one of the two major reasons I started with a blank slate for MAS.  The CUSTOM and other compound special variable systems are extremely cumbersome, but that's your only option in RPM if you want to do anything other than a basic prop.

**Comment by theonegalen** [comment]
I've gotten through the battery gauge finished so far, including testing. I'll try to have have the Classic Pack done this week.

**Comment by MOARdV** [comment]
> I've gotten through the battery gauge finished so far, including testing. I'll try to have have the Classic Pack done this week.

That's good progress.  I've been muddling through the JSI Basic MFD and picking off some of the RPM features that I had overlooked.  I hope to have that done this week, which will let me do a MAS release that officially supports the Stock RPM props.  Then I'll be back to ASET.

**Comment by MOARdV** [comment]
I found a couple of mystery switches - swTumbleERSHUD01 and swTumbleERSHUD02.  They both toggle persistent values (TurnERSHUD01 and TurnERSHUD02), but I don't see either of those persistents used anywhere in the ASET props.  -01 has a "ORIENTATION" caption below it, and -02 has a "COMPASS" caption.  The lighting for the caption on both is tied to whether landing gear are deployed on not.

**Comment by JonnyOThan** [comment]
> I found a couple of mystery switches - swTumbleERSHUD01 and swTumbleERSHUD02. They both toggle persistent values (TurnERSHUD01 and TurnERSHUD02), but I don't see either of those persistents used anywhere in the ASET props. -01 has a "ORIENTATION" caption below it, and -02 has a "COMPASS" caption. The lighting for the caption on both is tied to whether landing gear are deployed on not.

Probably related to https://forum.kerbalspaceprogram.com/index.php?/topic/85645-173-exploration-rover-system-by-aset-v03-040819/

**Comment by MOARdV** [comment]
Yeah, those props look like they're controls for the ERS Pitch/Roll Display in that mod (PitchRollDisplay.cfg)

**Comment by sovetskysoyuz** [comment]
Some of these are included in #368
